### PR TITLE
Fixes for Example Architecture

### PR DESCRIPTION
**Pull request by jnfoster**
Addresses mistakes noted in #340

### CODE REVIEW

**Comment by jnfoster** [comment]
@tomrodeheffer can you take a look also?

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -5585,48 +5585,43 @@ control EPipe<T, IH, OH>(in IH inputHeaders,
...
-
-Note that this architecture models a target switch that contains two
+  chained in the `Egress` package.  
+- The `Ingress.IPipe` is connected to the `Egress.EPipe`, because the
+  first outputs a value of type `T`, which is an input to the
+  second. Note that the the occurrences of the type variable `T` are
+  instantiated with the same type in `Switch`. In contrast, note that
+  the `Ingress` type `IH` and the `Egress` type `IH` may be
+  different. To force them to be the same type, we could instead
+  declare `IH` and `OH` at the switch level: `package Switch<IH, OH, T>`.
```
Comment (cc10512) [inline p4-16/spec/P4-16-spec.mdk]: you also have to change the argument signatures from `Ingress<T,_,_> ingress, Egress<T,_,_> egress`, to `Ingress<T,IH,OH> ingress, Ingress<T,IH,OH> egress`, as it is not clear how the defaults will propagate to arguments.

Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: The compiler will fill these in when Switch itself is instantiated.

Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: Just to confirm, the compiler accepts this program...
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T>(Ingress<T, _, _> ingress, Egress<T, _, _> egress);
```
If I wanted to make all of the type variables explicit, I'd have to parameterize `Switch` on additional type variables:
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
    
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T,H1,H2,H3,H4>(Ingress<T, H1, H2> ingress, Egress<T, H3, H4> egress);
```
Note that filling in `IH` and `OH` for the `_`s doesn't compile -- first, those type variables are unbound, and second, that would equate two types that are intended to be independent (per the comments above).

Comment (cc10512) [inline p4-16/spec/P4-16-spec.mdk]: so then where does this leave your statement that you can pull the IH and OH into the package decl: ```package Switch<IH, OH, T>```?

Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: Sorry, the second example above was to preserve the same (independent) design. 

To force them to be the same, we could write this:
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
    
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T,IH,OH>(Ingress<T, IH,OH> ingress, Egress<T, IH,OH> egress);
```

Comment (cc10512) [inline p4-16/spec/P4-16-spec.mdk]: Yes, that was what I thought and tried to express in the initial comment :)

Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: Now I'm getting confused. I think both are legal programs but they mean different things... Let's just sort it out in person tomorrow?

**Comment by mbudiu-vmw** [comment]
Please add these programs as tests for the compiler. I expected that the one in the text already exists.

File: p4-16/spec/P4-16-spec.mdk (hunk 1 above)
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: Done.

**Comment by jnfoster** [comment]
@cc10512 I added the complete statement for the example, as requested.

@mbudiu-vmw I added tests for p4c https://github.com/p4lang/p4c/pull/777

I think we're good to merge, yes?

**Pull request merged**